executor/analyze.go

```diff
@@ -502,16 +518,453 @@ type AnalyzeFastExec struct {
 	idxsInfo        []*model.IndexInfo
 	concurrency     int
 	maxNumBuckets   uint64
-	table           table.Table
 	cache           *tikv.RegionCache
 	wg              *sync.WaitGroup
 	sampLocs        chan *tikv.KeyLocation
-	sampLocRowCount uint64
-	tasks           chan *AnalyzeFastTask
+	rowCount        uint64
+	sampCursor      int32
+	sampTasks       []*AnalyzeFastTask
 	scanTasks       []*tikv.KeyLocation
 }
 
+func (e *AnalyzeFastExec) getSampRegionsRowCount(bo *tikv.Backoffer, needRebuild *bool, err *error, sampTasks *[]*AnalyzeFastTask) {
```

Comment:
Why we use pointer here for `error` and `bool`?

Reply:
This function is called by goroutine. Using pointer may easily control.

Reply:
can we use another function signature? for example:
```go
func (e *AnalyzeFastExec) getSampRegionsRowCount(bo *tikv.Backoffer) (needRebuild bool, sampTasks []*AnalyzeFastTask, err error) {

```

Reply:
Goroutine function cannot return values?

Reply:
how about wrapping these values as a struct, send the result to a channel. for example:
```go
// the struct name maybe not suitable.
type sampleTask struct {
	rebuild bool
	sampleTasks []*AnalyzeFastTask
	err error
}
```

in `buildSampTask()`:
```go
resultCh := make(chan sampleTask, e.concurrency)
```

in `getSampRegionsRowCount()`, after the routine is finished, send the result to the channel:
```go
result := sampleTask{/* fill values */ }
resultCh <- result
```

in `buildSampTask()`, read from the `resultCh` to get a finished task:
```go
result := <- resultCh
```